### PR TITLE
Check for BOM when converting

### LINKED ISSUE #263: Support for UTF-8-BOM
Reviewing #208 and #40, it looks like `readr` should support this, but the CRANed 0.1.1 is failing to handle a UTF-8-BOM CSV. If I supply `fileEncoding = UTF-8-BOM` to `read.csv` in base, this file comes in cleanly. Should `readr` be able to handle this?

This file is a MS SQL Server CSV export so...Windows. :wink:

### PR DESCRIPTION
**Pull request by jimhester**
This works for the test case, the main issue is it really only needs to check for this at the beginning of the file (if it is a UTF-8 file). I am still trying to figure out the best place to put this logic, suggestions welcome!

Fixes #263

### CODE REVIEW

**Comment by hadley** [comment]
Maybe it should be done in the constructor of the sources?

And maybe pull out into `hasUtf8Bom()` function?

**Comment by jimhester** [comment]
ˇI put the logic in a `skipBom()` function which is called from the three Source constructors.

Let me know if you think there is anything else we need to change.

File: src/Source.h [hunk 1]
```diff
@@ -47,6 +47,47 @@ class Source {
     return cur;
   }
 
+  static const char* skipBom(const char* begin) {
+    // Skip any unicode byte order marks
+    switch(begin[0]) {
+      // UTF-32BE
+      case '\x00':
+        if (begin[1] == '\x00' && begin[2] == '\xFE' && begin[3] == '\xFF') {
```
Comment (hadley) [inline src/Source.h]: I think this is a little risky because you might advance past the end of the string

Comment (hadley) [inline src/Source.h]: Otherwise looks good

Comment (hadley) [inline src/Source.h]: Just remembered I faced this problem else where, and the answer is `boost::starts_with`. See e.g. `inComment()`:

```cpp
  static bool inComment(const char* cur, const char* end, const std::string& comment) {
    boost::iterator_range<const char*> haystack(cur, end);
    return boost::starts_with(haystack, comment);
  }
```

Comment (hadley) [inline src/Source.h]: Would also be good to include a comment as to the sources of these byte sequences

Comment (jimhester) [inline src/Source.h]: Using an interator_range will likely have more overhead than doing the comparisons directly, the main benefit of the current code is each byte is compared only once, using any ranged compare would require you to compare the first few bytes more than once for each comparison.

I can make the current safe by passing the end ptr and verifying the length is at least (2 or 4) prior to comparison.

Comment (hadley) [inline src/Source.h]: It's up to you - I don't think this will make much difference to performance either way ;)

**Comment by jimhester** [comment]
You are right we are only calling this once per file. I switched to using `boost::starts_with()` for brevity/readability.

**Comment by hadley** [comment]
LGTM

**Pull request merged**